### ISSUE TITLE
Construct and spawn entity events when broken block dropping item

### ISSUE THREAD

**Issue by Lignium** [SpongePowered/Sponge#2144]
**I am currently running**

- SpongeVanilla version: spongevanilla-1.12.2-7.1.4
- Java version: OpenJDK 1.8.0_181
- Operating System: Kubuntu 18.04
- Plugins/Mods: Only one experimental plugin

**Issue Description**

Hi, SpongePowered team!

I conducted a very large analysis of events data which occurs when a broken block dropping an item. In total, I checked 11 different ways to break a flower pot and other blocks such as cactus and reeds.

Full log with my marks is located here: [ConstructEntityEvent and SpawnEntityEvent.pdf](https://github.com/SpongePowered/SpongeCommon/files/2699353/ConstructEntityEvent.and.SpawnEntityEvent.pdf)

Many mistakes are repeated, so I can draw some conclusions:
- Incorrect the `targetEntity` field in `ConstructEntityEvent.Post`
- Incorrect the `location` data in the `transform` field in `ConstructEntityEvent.Post`
- Firing `SpawnEntityEvent` instead of `DropItemEvent.Destruct`
- Forced the spawn type in context to `SpawnTypes.BLOCK_SPAWNING` in `SpawnEntityEvent` and `DropItemEvent.Destruct`

But again, for more information, carefully look at the log. Thanks!

P.S. Sorry for my English! Maybe I was wrong somewhere, but I think I wrote it clearly. :)

**Comment by gabizou** [comment]
Perhaps it would be beneficial as well to include stack traces to the already detailed pdf. The legend is helpful, however, I'm seeing some results of various areas that are phase state dependent, such that debugging would be aided by having said stack traces.

A lot of what you're seeing are indeed bugs, as the data is being predetermined by the [phase state that is handling said events during post process](https://github.com/SpongePowered/SpongeCommon/blob/stable-7/src/main/java/org/spongepowered/common/event/tracking/phase/block/BlockDropItemsPhaseState.java#L95-L117), and determined when the [phase state is delegated to "what is happening"](https://github.com/SpongePowered/SpongeCommon/blob/stable-7/src/main/java/org/spongepowered/common/event/tracking/phase/block/BlockDropItemsPhaseState.java#L53-L57). It's possible to fix these issues provided with your pdf data and if stack traces are provided.

**Comment by Lignium** [comment]
@gabizou, I understand correctly, do I need to run `Thread.currentThread().getStackTrace()` code for all 11 cases?

**Comment by gabizou** [comment]
Yes.

**Comment by Lignium** [comment]
Updated PDF... A month later :D

**Comment by gabizou** [comment]
So, not to dissuade that this issue isn't being worked on, I've started to actively add [test plugins](
https://github.com/SpongePowered/SpongeCommon/commit/679a1abd1154a2ea2f0266fac55e4bcda3882c13)  to target the tests outlined by this issue (validating and verifying the causes and context and position information that will be affecting construct.pre events etc). As soon as I can verify that the erroneous or missing data is corrected with these test plugins, I'll be closing the issue. As it stands, having the stack traces and causes/contexts printed out is a big help since it helps identify what and where these issues are being brought up in the implementation (and helps identify where the erroneous information is coming from by proxy of the PhaseTracker).

**Comment by Lignium** [comment]
Accidentally pressed the close button, damn)

**Comment by Lignium** [comment]
More than a year and a half later, this report is out of date. Stack traces are no longer relevant. The code was changed, as have the events. Given that I now understand more how the game works, Mixin and Sponge than then, I tried to figure it out myself. However, I almost did not succeed. I found some points in the code that are involved in the formation of these events, but this did not go further. PhaseTracker drives me crazy with its complexity. Without outside help, I definitely can’t do it. Perhaps this issue should be closed and created a new one with relevant data, but first describe one case, and not a dozen at once?
